<a href="https://colab.research.google.com/github/SachinScaler/Apr24_TimeSeries_and_RecSys/blob/main/ML_Recommender_Systems_1%7CLecture.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **Problem Statement**
Suppose you are a Data Scientist at a retail store like Walmart/DMart/RelianceSmartStore.
The company wants to increase sales and customer satisfaction through their purchase journey.
Your task is to use data to determine the products that are often purchased together:
- Retailers can optimize product placement
- Offer special deals
- create new product bundles to encourage further sales of these combinations

In [1]:
#downloading CSV Data from drive
id = "1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v"
print("https://drive.google.com/uc?export=download&id=" + id)
!wget "https://drive.google.com/uc?export=download&id=1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v" -O Online_Retail.csv

https://drive.google.com/uc?export=download&id=1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v
--2024-04-30 16:58:14--  https://drive.google.com/uc?export=download&id=1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v
Resolving drive.google.com (drive.google.com)... 173.194.203.100, 173.194.203.138, 173.194.203.101, ...
Connecting to drive.google.com (drive.google.com)|173.194.203.100|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://drive.usercontent.google.com/download?id=1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v&export=download [following]
--2024-04-30 16:58:14--  https://drive.usercontent.google.com/download?id=1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v&export=download
Resolving drive.usercontent.google.com (drive.usercontent.google.com)... 142.250.107.132, 2607:f8b0:400e:c08::84
Connecting to drive.usercontent.google.com (drive.usercontent.google.com)|142.250.107.132|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 45005763 (43M) [application/octet-stream]
Saving

In [2]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

df = pd.read_csv("Online_Retail.csv")
df.head(20)

InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5     536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6     536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7     536366     22633               HAND WARMER UNION JACK         6   
8     536366     22632            HAND WARMER RED POLKA DOT         6   
9     536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   
10    536367     22745           POPPY'S PLAYHOUSE BEDROOM          6   
11    536367     22748            POPPY'S PLAYHOUSE KITCHEN         6   
12    536367     22749    FELTCRAFT PRINCESS CHARLOTTE DOLL         8   
13    536367     22310              IVORY KNITTED MUG COSY          6   
14    536367     84969   BOX OF 6 ASSORTED COLOUR TEASPOONS         6   
15    536367     22623        BOX OF VINTAGE JIGSAW BLOCKS          3   
16    536367     22622       BOX OF VINTAGE ALPHABET BLOCKS         2   
17    536367     21754             HOME BUILDING BLOCK WORD         3   
18    536367     21755             LOVE BUILDING BLOCK WORD         3   
19    536367     21777          RECIPE BOX WITH METAL HEART         4   

      InvoiceDate  UnitPrice  CustomerID         Country  
0   01/12/10 8:26       2.55     17850.0  United Kingdom  
1   01/12/10 8:26       3.39     17850.0  United Kingdom  
2   01/12/10 8:26       2.75     17850.0  United Kingdom  
3   01/12/10 8:26       3.39     17850.0  United Kingdom  
4   01/12/10 8:26       3.39     17850.0  United Kingdom  
5   01/12/10 8:26       7.65     17850.0  United Kingdom  
6   01/12/10 8:26       4.25     17850.0  United Kingdom  
7   01/12/10 8:28       1.85     17850.0  United Kingdom  
8   01/12/10 8:28       1.85     17850.0  United Kingdom  
9   01/12/10 8:34       1.69     13047.0  United Kingdom  
10  01/12/10 8:34       2.10     13047.0  United Kingdom  
11  01/12/10 8:34       2.10     13047.0  United Kingdom  
12  01/12/10 8:34       3.75     13047.0  United Kingdom  
13  01/12/10 8:34       1.65     13047.0  United Kingdom  
14  01/12/10 8:34       4.25     13047.0  United Kingdom  
15  01/12/10 8:34       4.95     13047.0  United Kingdom  
16  01/12/10 8:34       9.95     13047.0  United Kingdom  
17  01/12/10 8:34       5.95     13047.0  United Kingdom  
18  01/12/10 8:34       5.95     13047.0  United Kingdom  
19  01/12/10 8:34       7.95     13047.0  United Kingdom

In [3]:
df.shape

(541909, 8)

In [4]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [5]:
df['Description'].nunique()

4223

Goal: from item level to transaction level

**Q. Why are some values negative under the Quantity and UnitPrice columns?**



### **Data Preprocessing**

In [6]:
print('Number of Unique Invoice numbers: {cnt}'.format(cnt=df.InvoiceNo.nunique()))

Number of Unique Invoice numbers: 25900


In [7]:
print('Number of Unique Customer IDs: {cnt}'.format(cnt=df.CustomerID.nunique()))

Number of Unique Customer IDs: 4372



**Let's get rid of all rows where quantity is < 0.**


In [8]:
df = df[df['Quantity']>=0]
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 531285 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    531285 non-null  object 
 1   StockCode    531285 non-null  object 
 2   Description  530693 non-null  object 
 3   Quantity     531285 non-null  int64  
 4   InvoiceDate  531285 non-null  object 
 5   UnitPrice    531285 non-null  float64
 6   CustomerID   397924 non-null  float64
 7   Country      531285 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 36.5+ MB


In [9]:
df = df[df['UnitPrice']>=0]
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 531283 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    531283 non-null  object 
 1   StockCode    531283 non-null  object 
 2   Description  530691 non-null  object 
 3   Quantity     531283 non-null  int64  
 4   InvoiceDate  531283 non-null  object 
 5   UnitPrice    531283 non-null  float64
 6   CustomerID   397924 non-null  float64
 7   Country      531283 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 36.5+ MB


In [10]:
df.shape

(531283, 8)

In [11]:
df[df['InvoiceNo'].str.contains('C')]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

**Let’s drop the rows that don’t have invoice numbers and remove the credit transactions (those with invoice numbers containing C).**

In [12]:
df.dropna(axis=0, subset=['InvoiceNo'],inplace=True)
df['InvoiceNo']=df['InvoiceNo'].astype('str')
df = df[~df['InvoiceNo'].str.contains('C')]
df.shape

(531283, 8)

In [13]:
df['Country'].value_counts()

Country
United Kingdom          486284
Germany                   9042
France                    8408
EIRE                      7894
Spain                     2485
Netherlands               2363
Belgium                   2031
Switzerland               1967
Portugal                  1501
Australia                 1185
Norway                    1072
Italy                      758
Channel Islands            748
Finland                    685
Cyprus                     614
Sweden                     451
Unspecified                446
Austria                    398
Denmark                    380
Poland                     330
Japan                      321
Israel                     295
Hong Kong                  284
Singapore                  222
Iceland                    182
USA                        179
Canada                     151
Greece                     145
Malta                      112
United Arab Emirates        68
European Community          60
RSA                         58


- Since most our data is from UK, let's just consider that and drop the rest.



In [14]:
data = df[df['Country'] =="United Kingdom"]
data.shape

(486284, 8)

In [15]:
data.describe()

Quantity      UnitPrice     CustomerID
count  486284.000000  486284.000000  354345.000000
mean        9.734676       3.840488   15552.436219
std       163.262763      34.636460    1594.546025
min         1.000000       0.000000   12346.000000
25%         1.000000       1.250000   14194.000000
50%         3.000000       2.100000   15522.000000
75%        10.000000       4.130000   16931.000000
max     80995.000000   13541.330000   18287.000000

In [16]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

     InvoiceDate  UnitPrice  CustomerID         Country  
0  01/12/10 8:26       2.55     17850.0  United Kingdom  
1  01/12/10 8:26       3.39     17850.0  United Kingdom  
2  01/12/10 8:26       2.75     17850.0  United Kingdom  
3  01/12/10 8:26       3.39     17850.0  United Kingdom  
4  01/12/10 8:26       3.39     17850.0  United Kingdom

### **Normalization:** Let's Normalize the Data and convert it into Sparse Matrix

In [17]:
data= data.groupby(['InvoiceNo', 'Description'])['Quantity'].sum()
data.head(20)

InvoiceNo  Description                        
536365     CREAM CUPID HEARTS COAT HANGER          8
           GLASS STAR FROSTED T-LIGHT HOLDER       6
           KNITTED UNION FLAG HOT WATER BOTTLE     6
           RED WOOLLY HOTTIE WHITE HEART.          6
           SET 7 BABUSHKA NESTING BOXES            2
           WHITE HANGING HEART T-LIGHT HOLDER      6
           WHITE METAL LANTERN                     6
536366     HAND WARMER RED POLKA DOT               6
           HAND WARMER UNION JACK                  6
536367     ASSORTED COLOUR BIRD ORNAMENT          32
           BOX OF 6 ASSORTED COLOUR TEASPOONS      6
           BOX OF VINTAGE ALPHABET BLOCKS          2
           BOX OF VINTAGE JIGSAW BLOCKS            3
           DOORMAT NEW ENGLAND                     4
           FELTCRAFT PRINCESS CHARLOTTE DOLL       8
           HOME BUILDING BLOCK WORD                3
           IVORY KNITTED MUG COSY                  6
           LOVE BUILDING BLOCK WORD                3

In [18]:
data = data.unstack()
data

Description   4 PURPLE FLOCK DINNER CANDLES   50'S CHRISTMAS GIFT BAG LARGE  \
InvoiceNo                                                                     
536365                                  NaN                             NaN   
536366                                  NaN                             NaN   
536367                                  NaN                             NaN   
536368                                  NaN                             NaN   
536369                                  NaN                             NaN   
...                                     ...                             ...   
581583                                  NaN                             NaN   
581584                                  NaN                             NaN   
581585                                  NaN                             NaN   
581586                                  NaN                             NaN   
A563185                                 NaN                             NaN   

Description   DOLLY GIRL BEAKER   I LOVE LONDON MINI BACKPACK  \
InvoiceNo                                                       
536365                      NaN                           NaN   
536366                      NaN                           NaN   
536367                      NaN                           NaN   
536368                      NaN                           NaN   
536369                      NaN                           NaN   
...                         ...                           ...   
581583                      NaN                           NaN   
581584                      NaN                           NaN   
581585                      NaN                           NaN   
581586                      NaN                           NaN   
A563185                     NaN                           NaN   

Description   NINE DRAWER OFFICE TIDY   OVAL WALL MIRROR DIAMANTE   \
InvoiceNo                                                            
536365                            NaN                          NaN   
536366                            NaN                          NaN   
536367                            NaN                          NaN   
536368                            NaN                          NaN   
536369                            NaN                          NaN   
...                               ...                          ...   
581583                            NaN                          NaN   
581584                            NaN                          NaN   
581585                            NaN                          NaN   
581586                            NaN                          NaN   
A563185                           NaN                          NaN   

Description   RED SPOT GIFT BAG LARGE   SET 2 TEA TOWELS I LOVE LONDON   \
InvoiceNo                                                                 
536365                            NaN                               NaN   
536366                            NaN                               NaN   
536367                            NaN                               NaN   
536368                            NaN                               NaN   
536369                            NaN                               NaN   
...                               ...                               ...   
581583                            NaN                               NaN   
581584                            NaN                               NaN   
581585                            NaN                               NaN   
581586                            NaN                               NaN   
A563185                           NaN                               NaN   

Description   SPACEBOY BABY GIFT SET   TOADSTOOL BEDSIDE LIGHT   ...  \
InvoiceNo                                                        ...   
536365                           NaN                        NaN  ...   
536366                 

In [19]:
data.shape

(18192, 4058)

In [20]:
data = data.fillna(0)
data.head()

Description   4 PURPLE FLOCK DINNER CANDLES   50'S CHRISTMAS GIFT BAG LARGE  \
InvoiceNo                                                                     
536365                                  0.0                             0.0   
536366                                  0.0                             0.0   
536367                                  0.0                             0.0   
536368                                  0.0                             0.0   
536369                                  0.0                             0.0   

Description   DOLLY GIRL BEAKER   I LOVE LONDON MINI BACKPACK  \
InvoiceNo                                                       
536365                      0.0                           0.0   
536366                      0.0                           0.0   
536367                      0.0                           0.0   
536368                      0.0                           0.0   
536369                      0.0                           0.0   

Description   NINE DRAWER OFFICE TIDY   OVAL WALL MIRROR DIAMANTE   \
InvoiceNo                                                            
536365                            0.0                          0.0   
536366                            0.0                          0.0   
536367                            0.0                          0.0   
536368                            0.0                          0.0   
536369                            0.0                          0.0   

Description   RED SPOT GIFT BAG LARGE   SET 2 TEA TOWELS I LOVE LONDON   \
InvoiceNo                                                                 
536365                            0.0                               0.0   
536366                            0.0                               0.0   
536367                            0.0                               0.0   
536368                            0.0                               0.0   
536369                            0.0                               0.0   

Description   SPACEBOY BABY GIFT SET   TOADSTOOL BEDSIDE LIGHT   ...  \
InvoiceNo                                                        ...   
536365                           0.0                        0.0  ...   
536366                           0.0                        0.0  ...   
536367                           0.0                        0.0  ...   
536368                           0.0                        0.0  ...   
536369                           0.0                        0.0  ...   

Description  returned  taig adjust  test  to push order througha s stock was   \
InvoiceNo                                                                       
536365            0.0          0.0   0.0                                  0.0   
536366            0.0          0.0   0.0                                  0.0   
536367            0.0          0.0   0.0                                  0.0   
536368            0.0          0.0   0.0                                  0.0   
536369            0.0          0.0   0.0                                  0.0   

Description  website fixed  wrongly coded 20713  wrongly coded 23343  \
InvoiceNo                                                              
536365                 0.0                  0.0                  0.0   
536366                 0.0                  0.0                  0.0   
536367                 0.0                  0.0                  0.0   
536368                 0.0                  0.0                  0.0   
536369                 0.0                  0.0                  0.0   

Description  wrongly marked  wrongly marked 23343  \
InvoiceNo                                           
536365                  0.0                   0.0   
536366                  0.0                   0.0   
536367                  0.0                   0.0   
536368                  0.0                   0.0   
536369                  0.0                   0.0   

Description  wrongly so

#### **Set Boolean:**
**For any transaction, if the quantity of an item is >=1 then its encoded as 1 (bought) , else 0 (not bought)**

In [21]:
def encode_units(x):
    if x >= 1:
        return 1
    else:
        return 0

data = data.applymap(encode_units)
data

Description   4 PURPLE FLOCK DINNER CANDLES   50'S CHRISTMAS GIFT BAG LARGE  \
InvoiceNo                                                                     
536365                                    0                               0   
536366                                    0                               0   
536367                                    0                               0   
536368                                    0                               0   
536369                                    0                               0   
...                                     ...                             ...   
581583                                    0                               0   
581584                                    0                               0   
581585                                    0                               0   
581586                                    0                               0   
A563185                                   0                               0   

Description   DOLLY GIRL BEAKER   I LOVE LONDON MINI BACKPACK  \
InvoiceNo                                                       
536365                        0                             0   
536366                        0                             0   
536367                        0                             0   
536368                        0                             0   
536369                        0                             0   
...                         ...                           ...   
581583                        0                             0   
581584                        0                             0   
581585                        0                             0   
581586                        0                             0   
A563185                       0                             0   

Description   NINE DRAWER OFFICE TIDY   OVAL WALL MIRROR DIAMANTE   \
InvoiceNo                                                            
536365                              0                            0   
536366                              0                            0   
536367                              0                            0   
536368                              0                            0   
536369                              0                            0   
...                               ...                          ...   
581583                              0                            0   
581584                              0                            0   
581585                              0                            0   
581586                              0                            0   
A563185                             0                            0   

Description   RED SPOT GIFT BAG LARGE   SET 2 TEA TOWELS I LOVE LONDON   \
InvoiceNo                                                                 
536365                              0                                 0   
536366                              0                                 0   
536367                              0                                 0   
536368                              0                                 0   
536369                              0                                 0   
...                               ...                               ...   
581583                              0                                 0   
581584                              0                                 0   
581585                              0                                 0   
581586                              0                                 0   
A563185                             0                                 0   

Description   SPACEBOY BABY GIFT SET   TOADSTOOL BEDSIDE LIGHT   ...  \
InvoiceNo                                                        ...   
536365                             0                          0  ...   
536366                 

**Since our goal is to find frequently occuring items, let's get rid of all transactions where only one product is bought.**

We are going to uncover the association between 2 or more items that is bought according to historical data



In [22]:
data = data[(data > 0).sum(axis=1) >= 2]
data

Description   4 PURPLE FLOCK DINNER CANDLES   50'S CHRISTMAS GIFT BAG LARGE  \
InvoiceNo                                                                     
536365                                    0                               0   
536366                                    0                               0   
536367                                    0                               0   
536368                                    0                               0   
536372                                    0                               0   
...                                     ...                             ...   
581582                                    0                               0   
581583                                    0                               0   
581584                                    0                               0   
581585                                    0                               0   
581586                                    0                               0   

Description   DOLLY GIRL BEAKER   I LOVE LONDON MINI BACKPACK  \
InvoiceNo                                                       
536365                        0                             0   
536366                        0                             0   
536367                        0                             0   
536368                        0                             0   
536372                        0                             0   
...                         ...                           ...   
581582                        0                             0   
581583                        0                             0   
581584                        0                             0   
581585                        0                             0   
581586                        0                             0   

Description   NINE DRAWER OFFICE TIDY   OVAL WALL MIRROR DIAMANTE   \
InvoiceNo                                                            
536365                              0                            0   
536366                              0                            0   
536367                              0                            0   
536368                              0                            0   
536372                              0                            0   
...                               ...                          ...   
581582                              0                            0   
581583                              0                            0   
581584                              0                            0   
581585                              0                            0   
581586                              0                            0   

Description   RED SPOT GIFT BAG LARGE   SET 2 TEA TOWELS I LOVE LONDON   \
InvoiceNo                                                                 
536365                              0                                 0   
536366                              0                                 0   
536367                              0                                 0   
536368                              0                                 0   
536372                              0                                 0   
...                               ...                               ...   
581582                              0                                 0   
581583                              0                                 0   
581584                              0                                 0   
581585                              0                                 0   
581586                              0                                 0   

Description   SPACEBOY BABY GIFT SET   TOADSTOOL BEDSIDE LIGHT   ...  \
InvoiceNo                                                        ...   
536365                             0                          0  ...   
536366                 

### **Code Implementation: Apriori**

In [23]:
from mlxtend.frequent_patterns import apriori

In [24]:
frequent_itemsets_plus = apriori(data,
                                 min_support=0.03,
                                 use_colnames=True)
frequent_itemsets_plus

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                           itemsets
0    0.051696                           (6 RIBBONS RUSTIC CHARM)
1    0.035129                  (60 CAKE CASES VINTAGE CHRISTMAS)
2    0.045529                      (60 TEATIME FAIRY CAKE CASES)
3    0.033557                   (72 SWEETHEART FAIRY CAKE CASES)
4    0.052663                       (ALARM CLOCK BAKELIKE GREEN)
..        ...                                                ...
172  0.030292  (LUNCH BAG SPACEBOY DESIGN , LUNCH BAG RED RET...
173  0.031199  (LUNCH BAG SUKI DESIGN , LUNCH BAG RED RETROSPOT)
174  0.032529  (PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN ...
175  0.033013  (ROSES REGENCY TEACUP AND SAUCER , PINK REGENC...
176  0.031925  (WOODEN FRAME ANTIQUE WHITE , WOODEN PICTURE F...

[177 rows x 2 columns]

In [25]:
frequent_itemsets_plus = frequent_itemsets_plus.sort_values('support', ascending=False).reset_index(drop=True)
frequent_itemsets_plus.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                              itemsets
0  0.129875  (WHITE HANGING HEART T-LIGHT HOLDER)
1  0.116331             (JUMBO BAG RED RETROSPOT)
2  0.100671            (REGENCY CAKESTAND 3 TIER)
3  0.095471                       (PARTY BUNTING)
4  0.084165             (LUNCH BAG RED RETROSPOT)

In [26]:
frequent_itemsets_plus['length'] = frequent_itemsets_plus['itemsets'].apply(lambda x: len(x))

frequent_itemsets_plus

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                                           itemsets  length
0    0.129875               (WHITE HANGING HEART T-LIGHT HOLDER)       1
1    0.116331                          (JUMBO BAG RED RETROSPOT)       1
2    0.100671                         (REGENCY CAKESTAND 3 TIER)       1
3    0.095471                                    (PARTY BUNTING)       1
4    0.084165                          (LUNCH BAG RED RETROSPOT)       1
..        ...                                                ...     ...
172  0.030473                   (CHRISTMAS CRAFT LITTLE FRIENDS)       1
173  0.030473                          (CREAM HEART CARD HOLDER)       1
174  0.030413     (LUNCH BAG CARS BLUE, LUNCH BAG  BLACK SKULL.)       2
175  0.030292  (LUNCH BAG SPACEBOY DESIGN , LUNCH BAG RED RET...       2
176  0.030171  (LUNCH BAG SUKI DESIGN , LUNCH BAG  BLACK SKULL.)       2

[177 rows x 3 columns]

In [31]:
data.shape[0]*0.03

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


496.16999999999996

- Using apriori algorithm, we filter frequent itemsets by giving minimum support value of 3%
- Length is the number of items in the itemset
- Based on the support threshold of 3%, there are 177 itemsets that are considered as frequently bought
- Eg: White hanging Heart T-Light Holder is the most frequently bought items with the support value of 0.129875. i.e the item is bought 2148 times out of the whole transaction

In [32]:
int(16539*0.129875)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


2148

### **Association_rules**: Implementation


In [ ]:
from mlxtend.frequent_patterns import association_rules

association_rules=association_rules(frequent_itemsets_plus,
                                    metric='lift',
                                    min_threshold=1
                                    )
association_rules = association_rules.sort_values('lift', ascending=False).reset_index(drop=True)
association_rules.head(5)

### **Conclusions**
- We can see GREEN REGENCY TEACUP AND SAUCER and PINK REGENCY TEACUP AND SAUCER are the items that has the highest association each other since these two items has the highest lift value, i.e 14.6

- This tells us that 'PINK REGENCY TEACUP AND SAUCER' is 14.6 times more likely to be bought by the customers who buy 'GREEN REGENCY TEACUP AND SAUCER' compared to the default likelihood sale of 'PINK REGENCY TEACUP AND SAUCER’

- Since the antecedent support > consequent support, the rule that applies is {GREEN REGENCY TEACUP AND SAUCER -> PINK REGENCY TEACUP AND SAUCER}



- This means that a customer has a higher tendency to buy PINK REGENCY TEACUP AND SAUCER AFTER they buy GREEN REGENCY TEACUP AND SAUCER. Not the other way around

- The confidence level for the rule is 0.6177, which shows that out of all the transactions that contain “GREEN REGENCY TEACUP AND SAUCER”, ~62 % contain PINK REGENCY TEACUP AND SAUCER too

- This could very valuable information, because we are now aware which products should we put the discounts on. We could give discounts on PINK REGENCY TEACUP AND SAUCER if a customer buy GREEN REGENCY TEACUP AND SAUCER etc

Similarly, we can draw other business conclusions from these association rules, which will lead in increase in revenue

Q**. What Business Strategy can we follow based on these findings?**

- **Item Placements:** We could place these two products together for easy accessiblity and increase sales

- **Product Bundling:** We could bundle these two products and sell it together at a discounted price compared to each price combined

- **Customer Recommendation & Discounts:** We could place PINK REGENCY TEACUP AND SAUCER at the cashier, and every time a customer buys GREEN REGENCY TEACUP AND SAUCER, we could offer and recommend them to buy PINK REGENCY TEACUP AND SAUCER with a lower price

